tests/unit/test_ecies.cpp

```diff
@@ -0,0 +1,186 @@
+#include "mococrw/util.h"
+#include "mococrw/symmetric_crypto.h"
+
+using namespace mococrw;
+using namespace ::testing;
+
+using testing::Eq;
+
+class ECIESTests : public ::testing::Test
+{
+public:
+    void SetUp() override
+    {
+        ECIESEncryptionCtxBuilder encBuilder(secp384PublicKey);
+        ECIESDecryptionCtxBuilder decBuilder(secp384Key);
+
+        const SymmetricCipherMode mode = SymmetricCipherMode::CBC;
+        const SymmetricCipherKeySize keySize = SymmetricCipherKeySize::S_256;
+        const SymmetricCipherPadding padding = SymmetricCipherPadding::PKCS;
+        const std::vector<uint8_t> zeroIV = std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(mode));
+
+        kdf = std::shared_ptr<X963KDF>(new X963KDF(openssl::DigestTypes::SHA512));
+        encBuilder.setKDF(kdf);
+        decBuilder.setKDF(kdf);
+
+        auto macFunc = [](const std::vector<uint8_t> &key) {
+            std::unique_ptr<MessageAuthenticationCode> mac_ptr(
+                        new mococrw::HMAC(openssl::DigestTypes::SHA512, key));
+            return mac_ptr;
+        };
+        auto cipherFunc = [zeroIV](const std::vector<uint8_t> &key, const SymmetricCipherOperation& operation) {
+            std::unique_ptr<SymmetricCipherI> symCiphPtr;
+            auto builder = mococrw::AESCipherBuilder(mode, keySize, key).setIV(zeroIV);
+            builder.setPadding(padding);
+            if (operation == SymmetricCipherOperation::Decryption) {
+                symCiphPtr = builder.buildDecryptor();
+            } else {
+                symCiphPtr = builder.buildEncryptor();
+            }
+
+
+            return symCiphPtr;
+        };
+
+
+        encBuilder.setMacFactoryFunction(macFunc);
+        encBuilder.setMacCipherKeySize(512 / 8);
+
+        encBuilder.setSymmetricCipherFactoryFunction(cipherFunc);
+        encBuilder.setSymmetricCipherKeySize(256 / 8);
+
+        decBuilder.setMacFactoryFunction(macFunc);
+        decBuilder.setMacCipherKeySize(512 / 8);
+
+        decBuilder.setSymmetricCipherFactoryFunction(cipherFunc);
+        decBuilder.setSymmetricCipherKeySize(256 / 8);
+
+        encCtx = encBuilder.buildEncryptionCtx();
+        decCtx = decBuilder.buildDecryptionCtx();
+    }
+
+protected:
+    std::unique_ptr<ECIESEncryptionCtx> encCtx;
+    std::unique_ptr<ECIESDecryptionCtx> decCtx;
+    std::shared_ptr<X963KDF> kdf;
+    std::shared_ptr<mococrw::HMAC> mac;
+    AsymmetricKeypair secp384Key = AsymmetricKeypair::generate(ECCSpec(openssl::ellipticCurveNid::SECP_384r1));
+    AsymmetricPublicKey secp384PublicKey = secp384Key;
+
+};
+
+TEST_F(ECIESTests, testWithoutSalt)
+{
+    std::vector<uint8_t> testString = {'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!'};
+    encCtx->update(testString);
+    auto ciphertext = encCtx->finish();
+    auto mac = encCtx->getMAC();
+    auto ephKey = encCtx->getEphemeralKey();
+
+    decCtx->setMAC(mac);
+    decCtx->setEphemeralKey(ephKey);
+    decCtx->update(ciphertext);
```

Comment:
Not the order this would be done in during streaming decryption. Please use setEphemeralKey, followed by update, then setMAC, then finish.

Reply:
Done